### PR TITLE
feat(cluster_resources): add leases and service account

### PR DESCRIPTION
**Pull request by DexterYan**
## Description, Motivation and Context

- add leases and service accounts for [Cluster resources collector](https://troubleshoot.sh/docs/collect/cluster-resources/)
- add configmap and volume attachments

Fixes: 
#1157
#1217

## Checklist

- [x] New and existing tests pass locally with introduced changes.
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] The commit message(s) are informative and highlight any breaking changes
- [x] Any documentation required has been added/updated. For changes to https://troubleshoot.sh/ create a PR [here](https://github.com/replicatedhq/troubleshoot.sh/pulls)

## Does this PR introduce a breaking change?
- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by xavpaice** [comment]
#1157 also requests daemonsets - is that planned (or already included)?

**Comment by DexterYan** [comment]
> #1157 also requests daemonsets - is that planned (or already included)?

daemonsets has been included before this PR.  https://troubleshoot.sh/docs/collect/cluster-resources/#cluster-resourcesdaemonsetsnamespacenamejson

**Comment by CpuID** [comment]
@DexterYan @xavpaice thoughts on tests around this one...?

looks like we don't have a whole lot of tests around it now, there's some around namespaces but not around specific resource types...

might be a bit of a can of worms though :) dedicated task in itself to add test coverage around cluster resources as a whole?

**Comment by CpuID** [comment]
also it looks like https://troubleshoot.sh/docs/collect/cluster-resources/ lists the various resource types (by filename that the collector writes out), docs PR to append to that list?

interesting that daemonset's were already on the docs list... but this PR adds support for them? 🤷

**Comment by xavpaice** [comment]
In the spirit of campsite rules, I'd like to see if we can add unit tests for the individual collectors here (e.g. `func serviceAccounts()`), we would however need to mock out the k8s client to do that.

I'd also like to see how far we can go with the e2e test, making, say, a serviceAccount and then confirming that the file appears in the support bundle.

**Comment by DexterYan** [comment]
I have added e2e tests for checking those resources have been collected in our support bundle. For the unit test, I have added leases, service accounts, configmaps and volume attachments. I think I will create another PR to add more tests for existing resources

File: pkg/collect/cluster_resources_test.go [hunk 1]
```diff
@@ -1,14 +1,200 @@
...
+			client := testclient.NewSimpleClientset()
+			ctx := context.Background()
+			err := createConfigMaps(client, tt.configMapName, tt.namespaces)
+			require.NoError(t, err)
+
+			configMaps, _ := configMaps(ctx, client, tt.namespaces)
+			assert.Equal(t, len(tt.namespaces), len(configMaps))
+
+			for _, ns := range tt.namespaces {
+				assert.NotEmpty(t, configMaps[ns+".json"])
```
Comment (CpuID) [inline pkg/collect/cluster_resources_test.go:53]: do we want to check that the config map name exists in the JSON...? at least checking 1 element would give confidence its valid JSON

File: pkg/collect/cluster_resources_test.go [hunk 2]
```diff
@@ -1,14 +1,200 @@
...
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			client := testclient.NewSimpleClientset()
+			ctx := context.Background()
+			err := createTestVolumeAttachments(client, tt.leaseName)
+			require.NoError(t, err)
+
+			volumeAttachments, _ := volumeAttachments(ctx, client)
+			assert.NotEmpty(t, volumeAttachments)
```
Comment (CpuID) [inline pkg/collect/cluster_resources_test.go:106]: do we want to check that the volume attachment name exists in the JSON...? at least checking 1 element would give confidence its valid JSON

File: pkg/collect/cluster_resources_test.go [hunk 3]
```diff
@@ -1,14 +1,200 @@
...
+			client := testclient.NewSimpleClientset()
+			ctx := context.Background()
+			err := createTestLeases(client, tt.leaseName, tt.namespaces)
+			require.NoError(t, err)
+
+			leases, _ := leases(ctx, client, tt.namespaces)
+			assert.Equal(t, len(tt.namespaces), len(leases))
+
+			for _, ns := range tt.namespaces {
+				assert.NotEmpty(t, leases[ns+".json"])
```
Comment (CpuID) [inline pkg/collect/cluster_resources_test.go:166]: do we want to check that the lease name exists in the JSON...? at least checking 1 element would give confidence its valid JSON

File: pkg/collect/cluster_resources_test.go [hunk 4]
```diff
@@ -1,14 +1,200 @@
...
+			client := testclient.NewSimpleClientset()
+			ctx := context.Background()
+			err := createTestServiceAccounts(client, tt.serviceAccountName, tt.namespaces)
+			require.NoError(t, err)
+
+			servicesAccounts, _ := serviceAccounts(ctx, client, tt.namespaces)
+			assert.Equal(t, len(tt.namespaces), len(servicesAccounts))
+
+			for _, ns := range tt.namespaces {
+				assert.NotEmpty(t, servicesAccounts[ns+".json"])
```
Comment (CpuID) [inline pkg/collect/cluster_resources_test.go:229]: do we want to check that the service account name exists in the JSON...? at least checking 1 element would give confidence its valid JSON

**Review by CpuID — CHANGES_REQUESTED** [review]
looking very close. couple of comments for minor changes

also can we get a docs PR together? thanks :)

File: pkg/collect/cluster_resources_test.go (hunk 4 above)
Comment (DexterYan) [reply pkg/collect/cluster_resources_test.go:229]: Thank you! I have added json content check to check those elements have been written,

**Comment by DexterYan** [comment]
> looking very close. couple of comments for minor changes
> 
> also can we get a docs PR together? thanks :)

Documents PR is ready https://github.com/replicatedhq/troubleshoot.sh/pull/520

**Review by CpuID — APPROVED** [review]

**Pull request merged**